### PR TITLE
count_all_results() with $this->reset_select() 

### PR DESCRIPTION
**Pull request by yaoshanliang**
On line 1358 of file 'system/database/DB_query_builder.php'，function count_all_results() will call reset_select(), and this will reset the condition of select query.
However, in some case, we need to get the results as well as count. For example, in pagination, we need to count all the results, but only results of 0-20 using limit.
If we code as

    $count = $this->db->count_all_results();//get total numbers
    $this->db->limit(0, 20);//get first 20
    $data = $this->db->get()->result();//this will not work

So, I suggest to add a parameter with boolean value to judge whether to call reset_select().

### CODE REVIEW

File: system/database/DB_query_builder.php [hunk 1]
```diff
@@ -1355,7 +1355,7 @@ public function get($table = '', $limit = NULL, $offset = NULL)
 	 * @param	string
 	 * @return	int
 	 */
-	public function count_all_results($table = '')
+	public function count_all_results($table = '', $reset = true)
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: true -> TRUE

File: system/database/DB_query_builder.php [hunk 2]
```diff
@@ -1366,7 +1366,11 @@ public function count_all_results($table = '')
 		$result = ($this->qb_distinct === TRUE)
 			? $this->query($this->_count_string.$this->protect_identifiers('numrows')."\nFROM (\n".$this->_compile_select()."\n) CI_count_all_results")
 			: $this->query($this->_compile_select($this->_count_string.$this->protect_identifiers('numrows')));
-		$this->_reset_select();
+
+                if($reset)
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: if ($reset === TRUE)

**Comment by narfbg** [comment]
You're also using spaces for indentation, you must use tabs, as per our [styleguide](http://www.codeigniter.com/userguide3/general/styleguide.html).

And a changelog entry + documentation update is also necessary.

**Comment by yaoshanliang** [comment]
@narfbg Sorry, It's my first time to pull request.
And now I added changelog and documentation.

File: system/database/DB_query_builder.php [hunk 3]
```diff
@@ -1366,7 +1367,11 @@ public function count_all_results($table = '')
 		$result = ($this->qb_distinct === TRUE)
 			? $this->query($this->_count_string.$this->protect_identifiers('numrows')."\nFROM (\n".$this->_compile_select()."\n) CI_count_all_results")
 			: $this->query($this->_compile_select($this->_count_string.$this->protect_identifiers('numrows')));
-		$this->_reset_select();
+
+		if($reset === TRUE)
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: You're still missing a space between `if` and the parenthesis

File: system/database/DB_query_builder.php [hunk 4]
```diff
@@ -1366,7 +1367,11 @@ public function count_all_results($table = '')
...
+		{
+		    $this->_reset_select();
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: Here you have 2 tabs + 4 spaces, must be 3 tabs.

File: system/database/DB_query_builder.php [hunk 5]
```diff
@@ -1353,9 +1353,10 @@ public function get($table = '', $limit = NULL, $offset = NULL)
 	 * returned by an Query Builder query.
 	 *
 	 * @param	string
+	 * @param	bool	TRUE: resets QB values; FALSE: leave QB vaules alone
```
Comment (narfbg) [inline system/database/DB_query_builder.php]: The third column here should contain the variable name, not description ... better just remove the description, it's quite obvious.

File: user_guide_src/source/changelog.rst [hunk 6]
```diff
@@ -176,6 +176,7 @@ Release Date: Not Released
    -  Added Interbase/Firebird database support via the *ibase* driver.
    -  Added ODBC support for ``create_database()``, ``drop_database()`` and ``drop_table()`` in :doc:`Database Forge <database/forge>`.
    -  Added support to binding arrays as ``IN()`` sets in ``query()``.
+   -  Added an optional second parameter to ``count_all_results()``.
```
Comment (narfbg) [inline user_guide_src/source/changelog.rst]: This doesn't belong here and you're practically inserting 2 entries for 1 change.

File: user_guide_src/source/changelog.rst [hunk 7]
```diff
@@ -192,6 +193,7 @@ Release Date: Not Released
       - Methods ``insert_batch()`` and ``update_batch()`` now return an integer representing the number of rows affected by them.
       - Methods ``where()``, ``or_where()``, ``having()`` and ``or_having()`` now convert trailing  ``=`` and ``<>``,  ``!=`` SQL operators to ``IS NULL`` and ``IS NOT NULL`` respectively when the supplied comparison value is ``NULL``.
       - Added method chaining support to ``reset_query()``, ``start_cache()``, ``stop_cache()`` and ``flush_cache()``.
+      - Added an optional second parameter to ``count_all_results`` that allows leaving QB values alone.
```
Comment (narfbg) [inline user_guide_src/source/changelog.rst]: Should be:

     - Added an optional second to ``count_all_results()`` to disable resetting of QB values.

**Comment by narfbg** [comment]
By documentation update, I meant database/query_builder.rst update, which is still missing.

You don't need to apologize or something, don't worry ... everybody needs some directions the first time.

**Comment by yaoshanliang** [comment]
@narfbg Changes have been made, please have a look, many thanks.

**Comment by narfbg** [comment]
Still needs more additions & polishing but I guess I'll do them myself ...

**Pull request merged**